scripts/generateRegexes.js

```diff
@@ -0,0 +1,104 @@
+"use strict";
+
+const fs = require("fs");
+const path = require("path");
+const regenerate = require("regenerate");
+
+const cp = {
+  Mark: require("unicode-10.0.0/General_Category/Mark/code-points.js"),
+  JT: require("./Joining_Type"),
+  CombiningClassVirama: require("./Canonical_Combining_Class").Virama,
+
+  // https://tools.ietf.org/html/rfc5893#section-1.4
+  L: require("unicode-10.0.0/Bidi_Class/Left_To_Right/code-points.js"),
+  R: require("unicode-10.0.0/Bidi_Class/Right_To_Left/code-points.js"),
+  AL: require("unicode-10.0.0/Bidi_Class/Arabic_Letter/code-points.js"),
+  EN: require("unicode-10.0.0/Bidi_Class/European_Number/code-points.js"),
+  ES: require("unicode-10.0.0/Bidi_Class/European_Separator/code-points.js"),
+  ET: require("unicode-10.0.0/Bidi_Class/European_Terminator/code-points.js"),
+  AN: require("unicode-10.0.0/Bidi_Class/Arabic_Number/code-points.js"),
+  CS: require("unicode-10.0.0/Bidi_Class/Common_Separator/code-points.js"),
+  NSM: require("unicode-10.0.0/Bidi_Class/Nonspacing_Mark/code-points.js"),
+  BN: require("unicode-10.0.0/Bidi_Class/Boundary_Neutral/code-points.js"),
+  ON: require("unicode-10.0.0/Bidi_Class/Other_Neutral/code-points.js"),
+};
+
+const regexes = {
+  //////////////////////
+  // Validity criteria
+  // http://www.unicode.org/reports/tr46/#Validity_Criteria
+
+  // Step 5
+  combiningMarks: regenerate(cp.Mark),
+
+  /////////////
+  // CONTEXTJ
+  // https://tools.ietf.org/html/rfc5892#appendix-A
+
+  // A.1. ZWNJ, Rule 2 / A.2. ZWJ, Rule 2
+  combiningClassVirama: regenerate(cp.CombiningClassVirama),
+
+  // A.1. ZWNJ, Rule 3
+  validZWNJ: `(?:${regenerate().add(cp.JT.L).add(cp.JT.D)})(?:${regenerate(cp.JT.T)})*` +
+             "\\u200C" +
+             `(?:${regenerate(cp.JT.T)})*(?:${regenerate([...cp.JT.R, ...cp.JT.D])})`,
+
+  //////////////
+  // BIDI Rule
+  // https://tools.ietf.org/html/rfc5893#section-2
+
+  // Step 1
+  bidiS1LTR: regenerate().add(cp.L),
+  bidiS1RTL: regenerate().add(cp.R).add(cp.AL),
+
+  // Step 2
+  bidiS2: `^(?:${regenerate()
```

Comment:
Yep. The output from `regenerate()` has the format `[...]|\u....[...]|\u....[...]`. Without the grouping the meaning would change when we append the `*` operator.